flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/StreamSinkITCase.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.streaming.runtime;
+
+import org.apache.flink.api.common.RuntimeExecutionMode;
+import org.apache.flink.api.common.typeinfo.IntegerTypeInfo;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ExecutionOptions;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.runtime.operators.sink.TestSink;
+import org.apache.flink.streaming.util.FiniteTestSource;
+import org.apache.flink.test.util.AbstractTestBase;
+
+import org.junit.Test;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.joining;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+/**
+ * Integration test for {@link org.apache.flink.api.connector.sink.Sink} run time implementation.
+ */
+public class StreamSinkITCase extends AbstractTestBase {
+
+	static final List<Integer> SOURCE_DATA = Arrays.asList(
+			895, 127, 148, 161, 148, 662, 822, 491, 275, 122,
+			850, 630, 682, 765, 434, 970, 714, 795, 288, 422);
+
+	static final Queue<String> STREAMING_COMMIT_QUEUE = new ConcurrentLinkedQueue<>();
+
+	static final Queue<String> STREAMING_GLOBAL_COMMIT_QUEUE = new ConcurrentLinkedQueue<>();
+
+	static final Queue<String> BATCH_COMMIT_QUEUE = new ConcurrentLinkedQueue<>();
+
+	static final Queue<String> BATCH_GLOBAL_COMMIT_QUEUE = new ConcurrentLinkedQueue<>();
+
+	@Test
+	public void streamingExecutionMode() throws Exception {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.enableCheckpointing(100);
+		FiniteTestSource<Integer> source = new FiniteTestSource<>(SOURCE_DATA);
+
+		env.addSource(source, IntegerTypeInfo.INT_TYPE_INFO)
+				.addSink(TestSink
+						.newBuilder()
+						.setDefaultCommitter((Supplier<Queue<String>> & Serializable) () -> STREAMING_COMMIT_QUEUE)
+						.setGlobalCommitter((Supplier<Queue<String>> & Serializable) () -> STREAMING_GLOBAL_COMMIT_QUEUE)
+						.build());
+
+		env.execute();
+
+		final List<String> expectedCommittedData = SOURCE_DATA
+				.stream()
+				.map(x -> Tuple3.of(x, null, Long.MIN_VALUE).toString())
+				.collect(
+						Collectors.toList());
+
+		// source send data two times
+		expectedCommittedData.addAll(expectedCommittedData);
```

Comment:
Might remove this line since add itself might contains risks. We may direct use `.flatMap(x -> Stream.of(x, x))` to duplicate the elements in the streaming operator before.

Reply:
thanks. I have already done like this.